File: two-sums/two-sum.dart



//Two sum using dart.
//Runtime: 279ms Beats: 93.83%
//Memory: 144MB Beats: 26.24%
void main(){
    var sol = Solution();
    var nums = <int>[];
    nums.add(10);
    nums.add(5);
    nums.add(2);
    nums.add(8);
    print(sol.twoSum(nums, 18));


}

class Solution {
  List<int> twoSum(List<int> nums, int target) {
      var vals = new Map();
      for(int i = 0 ; i < nums.length ; ++i){
          int diff = target - nums[i];
          if(vals.containsKey(diff)){
              List<int> output = [vals[diff], i];
              return output;
          }
          else{
            vals[nums[i]] = i;
          }
      }
        List<int> output = [0 , 0];
        return output;
  }
}

File: two-sums/two-sum.js
//This is two sum hashtable solution using javascript.
//This algorithm has a worst case time complexity of O(n) where
//n is the number of values in the nums array.
//Time: 58ms Beats: 92.92%
//Memory: 42.8MB Beats: 39.35%


/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    const my_map = {};
    for(let i = 0 ; i < nums.length ; ++i){
        
        let diff = target - nums[i];
        if(my_map[diff] == null){
            my_map[nums[i]] = i; 
        }
        else{
            return [i , my_map[diff]];
        }
    }
};
